File: LowCode/UI/Codegen.fs
module LowCode.UI.Codegen

open System

open AST

open LowCode.AST
open LowCode.Codegen

type Options<'ast when 'ast :> IAST> =
    { CompactCode     : bool
    ; IndentLevel     : int
    ; ScriptGenerator : ICodegen<'ast> option
    }

module Options =
    let unit =
        { CompactCode     = false
        ; IndentLevel     = 2
        ; ScriptGenerator = None
        }
    
type GeneratorState<'ast when 'ast :> IAST> =
    { CurrentIndentLevel : int
    ; Options : Options<'ast>
    }

module GeneratorState =
    let unit =
        { CurrentIndentLevel = 0
        ; Options = Options.unit
        }

let nl = System.Environment.NewLine

let indent value = nl + String(' ', value)

let rec attributes : Attributes -> string =
    function
    | (name, value)::xs -> sprintf " %s=\"%s\"%s" name value (attributes xs)
    | []                -> ""

and elements generatorState =
    let opts = generatorState.Options
    let level = generatorState.CurrentIndentLevel
    let rec element =
        function
        | x::xs ->
            indent level + codegen generatorState x + elements generatorState xs
        | []    ->
            indent (level - opts.IndentLevel)
    in element
    
and tag generatorState name attrs elmnts =
    let opts = generatorState.Options
    let es = elements { generatorState with CurrentIndentLevel = generatorState.CurrentIndentLevel + opts.IndentLevel } elmnts
    in sprintf "<%s%s>%s</%s>" name (attributes attrs) es name
    
and codegen generatorState =
    let tag = tag generatorState
    let rec codegen =
        function
        | HTML ast -> tag "html" [] [ast]
        | Body asts -> tag "body" [] asts
        | H1 (attrs, str) -> tag "h1" attrs [text str]
        | P (attrs, str) -> tag "p" attrs [text str]
        | Script l ->
            match generatorState.Options.ScriptGenerator with
            | Some gen ->
                match LowCode.Logic.Codegen.codegen gen l with
                | Ok code -> tag "script" [] [text code]
                | Error error -> error
            | None -> tag "script" [] []
        | Text str -> str
    in codegen

File: src/LowCode/UI/Codegen.hs
module LowCode.UI.Codegen
    ( Attribute
    , Attributes
    , HtmlAST (..)
    , Options (..)
    , defaultOptions
    , GeneratorState (..)
    , defaultGeneratorState
    )  where

import           Control.Monad.Trans.Class
import           Control.Monad.Trans.Except
import           Control.Monad.Trans.State
import           Data.Either
import           Data.Functor.Identity
import           Text.Printf (printf)
import           Data.Text (Text)
import qualified Data.Text as T

import qualified LowCode.Codegen as C
import           LowCode.Emit

type HTMLCodegen = StateT GeneratorState (Except Text)

type Attribute = (Text, Text)
type Attributes = [Attribute]
type Name = Text

data HtmlAST
    = Tag Name Attributes [HtmlAST]
    | Text Text

instance C.Codegen HtmlAST where
    codegen ast = runIdentity $ runExceptT $ evalStateT (htmlCodegen ast) defaultGeneratorState

data Options = Options
    { compactCode     :: Bool
    , indentLevel     :: Int
    } deriving (Eq, Show)

defaultOptions :: Options
defaultOptions = Options False 2

data GeneratorState = GeneratorState
    { currentIndentLevel :: Int
    , options            :: Options
    } deriving (Eq, Show)

defaultGeneratorState :: GeneratorState
defaultGeneratorState = GeneratorState 0 defaultOptions

indent :: (Emit gen) => HTMLCodegen gen
indent = do
    indent <- currentIndentLevel <$> get
    pure $ emit $ T.replicate indent " "

withIndent :: HTMLCodegen gen -> HTMLCodegen gen
withIndent action = do
    indent <- indentLevel . options <$> get
    modify $ \st -> st { currentIndentLevel = currentIndentLevel st + indent }
    result <- action
    modify $ \st -> st { currentIndentLevel = currentIndentLevel st - indent }
    pure result

genAttributes :: (Emit gen, Monoid gen) => Attributes -> HTMLCodegen gen
genAttributes = \case
    (name, value) : xs -> do
        attributes <- genAttributes xs
        pure $ mconcat
            [ emit " "
            , emit name
            , emit "=\""
            , emit value
            , emit "\""
            , attributes
            ]
    [] -> pure $ emit ""

genElements
    :: (Emit gen, Monoid gen)
    => [HtmlAST]
    -> HTMLCodegen gen
genElements = \case
    x : xs -> do
        code <- withIndent (htmlCodegen x)
        elements <- genElements xs
        pure $ mconcat [code, elements]
    [] -> pure $ emit ""

htmlCodegen
    :: (Emit gen, Monoid gen)
    => HtmlAST
    -> HTMLCodegen gen
htmlCodegen = \case
    Tag name attributes asts -> do
        attributes <- genAttributes attributes
        elements <- genElements asts
        indent' <- indent
        pure $ mconcat
            [ indent'
            , emit "<"
            , emit name
            , attributes
            , emit ">\n"
            , elements
            , emit "\n"
            , indent'
            , emit "</"
            , emit name
            , emit ">\n"
            ]
    Text str -> do
        indent' <- indent
        pure $ mconcat [indent', emit str]
